ddtrace/sampler.py

```diff
@@ -396,6 +398,20 @@ def _pattern_matches(self, prop, pattern):
         # Exact match on the values
         return prop == pattern
 
+    @cachedmethod()
+    def _matches(self, key):
+        # type: (Tuple[Optional[str], str]) -> bool
+        service, name = key
+        return all(
```

Comment:
Could we optimize here by short circuiting? AFAICT this will always loop over every pattern

Reply:
well, we sort of do in the match pattern function, if no rule is set we return immediately `True`.

we could do that here instead, only call if the pattern is not `NO_RULE`, but /shrug not sure there is much of a difference there.

what do you think? how would you think to optimize further?

Reply:
I'm thinking to rewrite this as

```
for prop, pattern in [(service, self.service), (name, self.name)]:
    if not self._pattern_matches(prop, pattern):
        return False
else:
    return True
```